### PR TITLE
Allow `getproperty` for `PointerWrapper`s with special `struct`s

### PR DESCRIPTION
**Pull request by JoshuaLampert**
This addresses the issue from #72. I'm not 100% sure if this is the proper way to solve the issue since the `try ... catch` seems a bit unsafe to me. But at least the MWE from the issue (and tests) pass. What do you think?

### CODE REVIEW

**Comment by sloede** [comment]
IMHO this is not the right way to go about this, since it uses exception handling for control flow.

If this (or something similar) is considered the right way to go about it, we might have to use `@generated` functions to avoid the runtime overhead of figuring out the type information.

**Comment by Gnimuc** [comment]
```
for ty = (:p4est_quadrant, :p4est_etc)
    @eval Base.getproperty(pw::PointerWrapper{$ty}, name::Symbol) = PointerWrapper(Base.getproperty(pw.pointer, name))
end
```

**Review by ranocha — CHANGES_REQUESTED** [review]
I agree with @sloede that using exception handling here does not feel good to me. Could you please benchmark it to see whether it introduces an efficiency penalty?
Alternatively, we might investigate falling back to `getproperty` if `findfirst` returns `nothing` - we should of course check whether it's safe and efficient to do so.

**Comment by JoshuaLampert** [comment]
I'm personally not a great fan of explicitly listing all of these structs manually. When we add more of these structs or some struct, for whatever reason, is renamed we would have to adapt the list. A solution without an explicit list of struct names feels more dynamic to me.

**Comment by JoshuaLampert** [comment]
@ranocha, yes, I can do that. Bu later because I am at my phone now.

**Comment by Gnimuc** [comment]
you could use Clang.jl to emit that list automatically.

**Comment by ranocha** [comment]
I guess the idea is to look for a solution that works not only for the types provided by P4est.jl at the moment. If we can find a way to make it work nicely (in an efficient and type-stable way) with general types, we would prefer that. If this turns out to be impossible for some reason, we can look into the option to hard-code everything.

File: src/pointerwrappers.jl [hunk 1]
```diff
@@ -58,10 +58,12 @@ Base.propertynames(::PointerWrapper{T}) where T = fieldnames(T)
...
   if isnothing(i)
-    error("type $(string(T)) has no field $name")
+    # For some `struct`s, `fieldnames` gives `data` and not the actual field names, but we can use `Base.getproperty` for pointers,
+    # see https://github.com/trixi-framework/P4est.jl/issues/72
+    return PointerWrapper(Base.getproperty(pointer(pw), name))
   end
 
-  PointerWrapper(fieldtype(T, i), pointer(pw) + fieldoffset(T, i))
+  return PointerWrapper(fieldtype(T, i), pointer(pw) + fieldoffset(T, i))
 end
```
Comment (sloede) [inline src/pointerwrappers.jl:67]: I wonder if we should try to use a `@generated` function, now that the logic is becoming more and more complicated here. Maybe something like this:
```julia
# Syntactic sugar: allows one to use `pw.fieldname` to get a PointerWrapper-wrapped pointer to `fieldname`
@generated function Base.getproperty(pw::PointerWrapper{T}, name::Symbol) where T
  i = findfirst(isequal(name), fieldnames(T))
  if isnothing(i)
    # For some `struct`s, `fieldnames` gives `data` and not the actual field names, but we can use `Base.getproperty` for pointers,
    # see https://github.com/trixi-framework/P4est.jl/issues/72
    return quote
      $(Expr(:meta, :inline))
      PointerWrapper(Base.getproperty(pointer(pw), name))
    end
  else
    return quote
      $(Expr(:meta, :inline))
      PointerWrapper($(fieldtype(T, i)), pointer(pw) + $(fieldoffset(T, i)))
    end
  end
end
```
But don't assume I got the macro syntax correct 😅

Comment (ranocha) [reply src/pointerwrappers.jl:67]: I think this needs some debugging. Note that you have only access to the types when using a `@generated` function, see https://docs.julialang.org/en/v1/manual/metaprogramming/#Generated-functions
Thus, I would expect that we can only see that `name` is a `Symbol` when calling such a `@generated` function - while we would like to use its value...

**Review by ranocha — APPROVED** [review]

File: src/pointerwrappers.jl (hunk 1 above)
Comment (sloede) [reply src/pointerwrappers.jl:67]: I've seen this used elsewhere before, where they encoded this in `Val{...}`s and then were actually able to access the type in `@generated` functions. However, this seems overkill here, so let's go down that road only if and when necessary.

File: src/pointerwrappers.jl [hunk 2]
```diff
@@ -58,10 +58,12 @@ Base.propertynames(::PointerWrapper{T}) where T = fieldnames(T)
 function Base.getproperty(pw::PointerWrapper{T}, name::Symbol) where T
   i = findfirst(isequal(name), fieldnames(T))
   if isnothing(i)
-    error("type $(string(T)) has no field $name")
+    # For some `struct`s, `fieldnames` gives `data` and not the actual field names, but we can use `Base.getproperty` for pointers,
+    # see https://github.com/trixi-framework/P4est.jl/issues/72
+    return PointerWrapper(Base.getproperty(pointer(pw), name))
```
[the remark this answers is not in the document]
Comment (ranocha) [reply src/pointerwrappers.jl:63]: Shouldn't it throw a very similar error already in this case?

[the remark this answers is not in the document]
Comment (JoshuaLampert) [reply src/pointerwrappers.jl:63]: Yes, it should. If I recall correctly I have also tested that.

[the remark this answers is not in the document]
Comment (JoshuaLampert) [reply src/pointerwrappers.jl:63]: I just doublechecked it. The current error message is quite clear.

[the remark this answers is not in the document]
Comment (sloede) [reply src/pointerwrappers.jl:63]: Sounds good, thanks for checking!

**Pull request merged**